### DISCUSSION TITLE
Avalonia's AOT problem

### DISCUSSION THREAD

**Discussion by WenLiHou** [AvaloniaUI/Avalonia#13321] [General]
When implementing multi-language switching in avalonia, a language resource class implements INotifyPropertyChanged, takes the instance of this class as the binding source, switches multiple languages before aot, everything is fine, but after aot, the switch has a problem, INotifyPropertyChanged no longer triggers. What's the problem?

**Comment by WenLiHou** [comment]
https://github.com/WenLiHou/avalonia-aot-problem/tree/main

**Comment by timunie** [comment]
For AOT you need to have everything trimming friendly. If your Binding is a ReflectionBinding, that will not work out of the box. Check docs about difference between CompiledBinding and ReflectionBinding. 

Also check docs about .NET avoiding trimming of reflection components. I believe you need a roots.xml file

**Reply by emmauss** [reply]
The required file is a rd.xml file.

**Comment by emmauss** [comment]
You would need to update your MarkupExtension to provide a CompiledBinding instead of a ReflectionBinding

**Reply by WenLiHou** [reply]
I tried, but still.
------------------------
`	<StackPanel>
		<TextBlock Text="{Binding HelloWorld}"></TextBlock>

		<TextBlock Text="{CompiledBinding Localization[String1]}"></TextBlock>
		<ComboBox  ItemsSource="{Binding Localization.AvailableCultures}"   SelectedItem="{Binding Localization.CurrentCulture}"/>

		<Button Command="{Binding ClickCommand}" >switch</Button>
	</StackPanel>`

**Reply by emmauss** [reply]
How did it look when using reflection binding

**Reply by WenLiHou** [reply]
When using Reflection Binding, the interface is not working properly with or without AOT.

sample code: https://github.com/WenLiHou/avalonia-aot-problem/tree/main

**Reply by emmauss** [reply]
You've set both `AvaloniaUseCompiledBindingsByDefault` in your project file and `x:DataType` in your main view. Avalonia will use CompiledBindings anywhere `Binding` is used. You are already using CompiledBinding, no need to explicitly specify `CompiledBinding` when binding. Though it's strange that `CompiledBinding` extension isn't working as it should

**Reply by WenLiHou** [reply]
Yes, the interface is not working properly and I am distressed by it

**Reply by stevemonaco** [reply]
Something is relying upon System.Linq.Expressions various codegen features and I'm not familiar enough with resx to know how it works. You should research to see if it relies on features that will break AOT. I suggest trying to start a new project, add the properties below to your csproj, and slowly start adding components until the warnings come into the mix. There's well over 30+, even after removing ReactiveUI (which also has a few warnings). It's a big mistake to go into an AOT project without having the various analyzers/warns setup correctly. I doubt there's an issue with `CompiledBinding` as the binding works as long as it doesn't go through the resource lookup.

```xml
<PropertyGroup Label="AOT-related">
  <PublishTrimmed>true</PublishTrimmed>
  <PublishAot>true</PublishAot>
  <IsTrimmable>true</IsTrimmable>
  <EnableTrimAnalyzer>true</EnableTrimAnalyzer>
  <EnableAotAnalyzer>true</EnableAotAnalyzer>
  <EnableSingleFileAnalyzer>true</EnableSingleFileAnalyzer>
  <TrimmerSingleWarn>false</TrimmerSingleWarn>
</PropertyGroup>
```

Some of the warnings in the Output:
```
3>Avalonia.Win32.WinRT.WinRTInspectable.GetIids(UInt64*,Guid**): 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.Interfaces' in call to 'System.Type.GetInterfaces()'. The return value of method 'System.Object.GetType()' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
3>Avalonia.Skia.Metal.SkiaMetalApi.CreateContext(IntPtr,IntPtr,GRContextOptions): Using member 'System.Runtime.InteropServices.Marshal.SizeOf(Object)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. Marshalling code for the object might not be available. Use the SizeOf<T> overload instead.
3>Avalonia.Skia.Metal.SkiaMetalApi.CreateContext(IntPtr,IntPtr,GRContextOptions): Using member 'System.Runtime.InteropServices.Marshal.StructureToPtr(Object,IntPtr,Boolean)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. Marshalling code for the object might not be available. Use the StructureToPtr<T> overload instead.
3>System.Linq.Expressions.Interpreter.NewArrayInitInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available.
3>System.Linq.Expressions.Interpreter.NewArrayInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32)' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available.
3>System.Linq.Expressions.Interpreter.NewArrayBoundsInstruction.Run(InterpretedFrame): Using member 'System.Array.CreateInstance(Type,Int32[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The code for an array of the specified type might not be available.
3>System.Dynamic.Utils.TypeUtils.GetNullableType(Type): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.
3>System.Linq.Expressions.Compiler.DelegateHelpers.TryMakeVBStyledCallSite(Type[]): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.
3>System.Linq.Expressions.Compiler.DelegateHelpers.GetActionType(Type[]): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.
3>System.Linq.Expressions.Compiler.DelegateHelpers.GetFuncType(Type[]): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.
3>System.Linq.Expressions.Expression.ApplyTypeArgs(MethodInfo,Type[]): Using member 'System.Reflection.MethodInfo.MakeGenericMethod(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.
3>System.Linq.Expressions.Expression.GetResultTypeOfShift(Type,Type): Using member 'System.Type.MakeGenericType(Type[])' which has 'RequiresDynamicCodeAttribute' can break functionality when AOT compiling. The native code for this instantiation might not be available at runtime.
```

**Reply by WenLiHou** [reply]
I tried, but found that the warnings of the original template project output are the same as the warnings of the current project output.

**Reply by stevemonaco** [reply]
Warnings imply that certain pieces are not compatible with trimming and/or AOT. So you need to narrow that down and alter your design to avoid incompatible dependencies. I'd suspect it's .NET resx support in general, as I've built AOT demos with the remaining features, AFAIK.

**Reply by WenLiHou** [reply]
You're right, it's really the resx file problem, I removed it and replaced it with something else just fine.

**Comment by Coloryr** [comment]
所有东西都要用CompiledBinding，我用CompiledBinding之后，AOT之后啥东西都有用了，不用CompiledBinding有些东西都不触发更新的

**Reply by WenLiHou** [reply]
我上面的示例代码就是全部用CompiledBinding的，在没有AOT之前一切正常，AOT之后切换中英文就不正常了。。

**Reply by Coloryr** [reply]
我试试我的项目，应该说不会的

**Reply by Coloryr** [reply]
确实不行，AOT就寄了

**Reply by WenLiHou** [reply]
我试了，确实是如上面老哥说的resx文件问题，这个不兼容AOT吧，我不用resx文件就ok了
